### PR TITLE
[WIP] Tim/dsviz/keras callback data merged with art sync

### PR DESCRIPTION
**Pull request by tssweeney**
This PR is a WIP and is expected to be landed after the DSViz + Datasets API improvements. However, doing a first pass helped shape the design of the aforementioned work. Here is the goal of this PR: lower the barrier of entry for users to start using DSViz. We are not, however trying to solve all problems with WandB callback, as the complexity increases, we expect users to begin using the Table/logging apis directly. However, in an ideal case here is what we unlock:

### Case 0: No wandb
```
x, y = generate_dataset_blackbox()
train_x, train_y, test_x, test_y = generate_training_data_blackbox(x, y)
model.fit(train_x, train_y, validation_data=(test_x, test_y))
```

### Case 1: Basic Prediction logging
In this case, we add a one-liner instructing the WandbCallback to log our model's predictions against the validation set. Since the validation data is going to be an ndarray-like matrix of values, we will attempt to guess the input, label, and output types. Users can specify them directly as well as specify class label names (this matches today's pattern). Additionally, if the validation set is the same as prior runs, a new artifact will not be created.

```
wandb.init()
x, y = generate_dataset_blackbox()
train_x, train_y, test_x, test_y = split_data(x, y)
model.fit(train_x, train_y, validation_data=(test_x, test_y), callbacks=[WandbCallback(
   log_validation = True
)])
```

This will result in the following graph over multiple runs (notice that the validation table is not run scoped, but each epoch is):
Screen Shot 2021-03-02 at 12 37 58 PM

and a validation table looks like this:
Screen Shot 2021-03-02 at 12 38 35 PM

https://wandb.ai/timssweeney/my-keras-integration_v2/artifacts/history_table/validation_results-2kedb01x/f30a99c9bf5445f76831/files/validation_results.joined-table.json

Disclaimer: the current commit requires the user to provide a data processor function rather than guessing the types like this:
```
data_converter=LambdaMediaTableConverter(
      lambda x: [wandb.Image(x)],
      lambda x, y: [CLASS_NAMES[y.tolist()[0]]],
      lambda x, o: [CLASS_NAMES[o.argmax()]] + o.tolist(),
      ["image"],
      ["true_class"],
      ["pred_class"] + ["{}_score".format(clss) for clss in CLASS_NAMES]
  )
```

### Case 2: Sharing Data between Runs and Users:
We also want to support consuming validation data produced by other runs. for example (this works in current commit):
```
wandb.init()
x, y = generate_dataset_blackbox()
train_x, train_y, test_x, test_y = split_data(x, y)
_ = model.fit(
    x_train,
    y_train,
    epochs=config.epochs,
    batch_size=config.batch_size,
    validation_data=WandbCallback.use_validation_data("validation_data-15m3dmst:v0"),
    callbacks=[WandbCallback(
        log_evaluation=True,
)
```
However, as you might be guessing, the user may very well mix training and validation data if they are not careful. Which is why we need to track the data one level up - at the more raw data level. **This is why I paused this PR to work on the DSViz/Table/Dataset APIs so that they can be leveraged to get past this.**

### Case 3: Wholesale data integration (Waiting for other workstream)
```
wandb.init()
dataset = wandb.Artifact("example_dataset")
table = dataset["training_data"] 
if table is None: # if the data has not been created, create it
   x, y = generate_dataset_blackbox()
   table = wandb.Table(column_data=[
      ("x", x, wandb.Image),
      ("y", y, wandb.Classes(CLASS_NAMES))
   ]) # Create a typed table based on columns of numpy arrays - if not provided, ID column defaults to index
   dataset["training_data"] = table
   dataset.save()
train, test = table.split([0.7])
_ = model.fit(
    train["x"].numpy(),
    train["y"].numpy(),
    validation_data=(test["x"].numpy(), test["y"].numpy()),
    callbacks=[WandbCallback(
        log_evaluation=True,
       validation_ids = test["_id"]
   )])
```
The above code will create a dataset once, provide numpy/columnar accessors, and properly join predictions on the underlying shuffled/split data without creating duplicate data tables.

Description
-----------

Testing
-------

How was this PR tested?

### CODE REVIEW

**Comment by tssweeney** [comment]
Close for newer PR: https://github.com/wandb/client/pull/1974
